mlir/lib/Dialect/MIOpen/utility/transformMapUtils.cpp

```diff
@@ -213,3 +225,171 @@ TransformOp mlir::miopen::reshapeBuffer(OpBuilder &b, Location loc,
+          }
+        }
+      }
+      result[lowerDims[0]] = maxLength;
+      break;
+    }
+    // Like `Embed`, but a bit simpler since we don't have to sort.
+    case TransformType::Unmerge: {
+      int64_t stride = 1;
+      Optional<int64_t> maxLength;
+      for (auto pair :
+           llvm::zip(llvm::reverse(upperDims), llvm::reverse(params))) {
+        uint32_t upperDim = std::get<0>(pair);
+        int64_t dimLength = std::get<1>(pair);
+        if (input[upperDim].hasValue()) {
+          if (!maxLength.hasValue())
+            maxLength = 1;
+          if (maxLength == stride) {
+            maxLength = maxLength.getValue() * input[upperDim].getValue();
+          } else {
+            break;
+          }
+        }
+        stride *= dimLength;
+      }
+      result[lowerDims[0]] = maxLength;
+      break;
+    }
+    // For merges, the input vectorization length is split among the output
+    // dimensions, with a dimension getting a vectorization length equal
+    // to the gcd of the remaining vectorization length and that dimension's
+    // length. The remaining length gets lowered by the full lenght of the
+    // output dimension so that Merge{3, 4} with a starting vectorization lenght
+    // of 6 gets the results [1, 2] and not [3, 2].
+    case TransformType::Merge:
+    case TransformType::Unfold: {
+      int64_t upperDim = upperDims[0];
+      if (!input[upperDim].hasValue()) {
+        break;
+      }
+      int64_t maxLen = input[upperDim].getValue();
+      for (auto pair :
+           llvm::zip(llvm::reverse(lowerDims), llvm::reverse(params))) {
+        uint32_t lowerDim = std::get<0>(pair);
+        int64_t lowerLen = std::get<1>(pair);
+        int64_t thisMaxLen = math_util::gcd(maxLen, lowerLen);
+        result[lowerDim] = thisMaxLen;
+        maxLen = std::max(maxLen / lowerLen, 1L);
+      }
+      break;
+    }
+    }
+  }
+  return result;
+}
+
+int64_t mlir::miopen::getMaxVectorization(ArrayAttr transforms, uint32_t dim,
+                                          int64_t len,
+                                          ArrayRef<int64_t> outputShape) {
+  if (transforms.empty())
+    return len;
+
+  VectorizationData maxLengths;
+  maxLengths.grow(transforms[0]
+                      .cast<TransformMapAttr>()
+                      .getMap()
+                      .getValue()
+                      .getNumInputs());
+  maxLengths[dim] = len;
+  for (auto transformMap : transforms.getAsRange<TransformMapAttr>()) {
+    maxLengths = propagateVectorizationInfo(transformMap, maxLengths);
+  }
+
+  int64_t result = 1;
+  int64_t stride = 1;
+  for (size_t i = 0, e = outputShape.size(); i < e; ++i) {
+    uint32_t dimension = (e - 1) - i;
+    int64_t length = outputShape[dimension];
+    if (maxLengths[dimension].hasValue()) {
+      int64_t maxLength = maxLengths[dimension].getValue();
+      if (result == stride) {
+        result *= maxLength;
+      } else {
+        break;
+      }
```

Comment:
The problem there is vectorization probes where the final dimension was held constant, for example. 
That is, suppose we have NHWK output, and we want to vectorize the GemmN -> (N x H x W) dimension.
We can't do this unless K = 1, but your simplied logic would allow vectors in W.